### PR TITLE
Supervisor Subject New/Create

### PR DESCRIPTION
**Pull request by shakil-fg**

### CODE REVIEW

**Comment by noor-framgia** [comment]
ok

**Comment by mdmahmudurrahman** [comment]
OK

File: app/controllers/supervisor/subjects_controller.rb [hunk 1]
```diff
@@ -15,9 +27,12 @@ def show
...
   end
 
-
   private
     def load_user
       @user = User.first #Incompatibility 1: Please see SpecificationNotes.md
     end
+
+    def subject_params
+      params.require(:subject).permit(:name, :description, tasks_attributes: [:id, :name, :description, :_destroy])
```
Comment (dieunb) [inline app/controllers/supervisor/subjects_controller.rb]: to long, max 80 chars

File: app/controllers/supervisor/subjects_controller.rb [hunk 2]
```diff
@@ -15,9 +27,12 @@ def show
   def edit
   end
 
-
   private
     def load_user
       @user = User.first #Incompatibility 1: Please see SpecificationNotes.md
     end
+
+    def subject_params
```
Comment (dieunb) [inline app/controllers/supervisor/subjects_controller.rb]: indent

File: app/models/subject.rb [hunk 3]
```diff
@@ -3,7 +3,9 @@ class Subject < ActiveRecord::Base
   has_many :courses, through: :course_subjects, autosave: true
   has_many :users, through: :courses
 
-  has_many :tasks
+  has_many :tasks, dependent: :destroy
 
-  accepts_nested_attributes_for :tasks
+  accepts_nested_attributes_for :tasks, reject_if: proc { |attributes| attributes['name'].blank? }, allow_destroy: true
```
Comment (dieunb) [inline app/models/subject.rb]: too long, use double quote pls

**Comment by manazirahsan** [comment]
okey

**Comment by mdmahmudurrahman** [comment]
Okey

File: app/models/subject.rb [hunk 4]
```diff
@@ -3,7 +3,11 @@ class Subject < ActiveRecord::Base
   has_many :courses, through: :course_subjects, autosave: true
   has_many :users, through: :courses
 
-  has_many :tasks
+  has_many :tasks, dependent: :destroy
 
-  accepts_nested_attributes_for :tasks
+  accepts_nested_attributes_for :tasks,
+    reject_if:
+      proc { |attributes| attributes['name'].blank? },allow_destroy: true
```
Comment (dieunb) [inline app/models/subject.rb]: double quote pls

File: app/views/shared/_error_messages.html.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<% if object.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(object.errors.count, "error") %>.
```
Comment (dieunb) [inline app/views/shared/_error_messages.html.erb]: i18n pls

File: app/views/supervisor/subjects/_tasks_form.html.erb [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<div class="jumbotron">
+  <%= f.label :name %>
+  <%= f.text_field :name, class: 'form-control' %>
+
+  <%= f.label :description %>
+  <%= f.text_area :description, class: 'form-control' %>
+
+  <%= link_to_remove_fields "Remove", f %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/_tasks_form.html.erb]: i18n

File: app/views/supervisor/subjects/new.html.erb [hunk 7]
```diff
@@ -1,2 +1,24 @@
...
+
+      <%= subjects_form_obj.label :description %>
+      <%= subjects_form_obj.text_area :description, class: 'form-control' %>
+
+      <h3><%= t :tasks %></h3>
+
+      <%= subjects_form_obj.fields_for :tasks do |tasks_form_obj| %>
+        <%= render "tasks_form", f: tasks_form_obj %>
+      <% end %>
+      <p><%= link_to_add_fields "Add tasks", subjects_form_obj, :tasks %></p>
```
Comment (dieunb) [inline app/views/supervisor/subjects/new.html.erb]: i18n pls

**Comment by mdmahmudurrahman** [comment]
Ok

File: app/views/supervisor/subjects/new.html.erb [hunk 8]
```diff
@@ -1,2 +1,24 @@
-<h1>Subjects#new</h1>
-<p>Find me in app/views/subjects/new.html.erb</p>
+<h2><%= t "create_subject" %></h2>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for [:supervisor, @subject] do |subjects_form_obj| %>
+      <%= render 'shared/error_messages', object: subjects_form_obj.object %>
```
Comment (dieunb) [inline app/views/supervisor/subjects/new.html.erb]: double quote

**Pull request merged**